pep-0639.rst

```diff
@@ -1,543 +1,2220 @@
+Finally, `guidance is established <Converting legacy metadata_>`_
+for tools handling and converting legacy license metadata to license
+expressions, to ensure the results are consistent, correct and unambiguous.
+
+Note that the guidance on errors and warnings is for tools' default behavior;
+they MAY operate more strictly if users explicitly configure them to do so,
+such as by a CLI flag or a configuration option.
+
+
+Core metadata
+-------------
+
+The `PyPA Core Metadata specification <#coremetadataspec_>`_ defines the names
+and semantics of each of the supported fields in the distribution metadata of
+Python distribution packages and installed projects.
+
+This PEP `adds <Add License-Expression field_>`_ the
+``License-Expression`` field,
+`adds <Add License-File field_>`_ the ``License-File`` field,
+`deprecates <Deprecate License field_>`_ the ``License`` field,
+and `deprecates <Deprecate license classifiers_>`_ the license classifiers
+in the ``Classifier`` field.
+
+The error and warning guidance in this section applies to build and
+publishing tools; end-user-facing install tools MAY be more lenient than
+mentioned here when encountering malformed metadata
+that does not conform to this specification.
+
+As it adds new fields, this PEP updates the core metadata to version 2.3.
+
+
+Add ``License-Expression`` field
+''''''''''''''''''''''''''''''''
+
+The ``License-Expression`` optional field is specified to contain a text string
+that is a valid SPDX license expression, as defined herein.
+
+Publishing tools SHOULD issue an informational warning if this field is
+missing, and MAY raise an error. Build tools MAY issue a similar warning,
+but MUST NOT raise an error.
 
-Over time, recommending the usage of these expressions will help Python package
-publishers improve the clarity of their license documentation to the benefit of
-package authors, consumers and redistributors.
+A license expression is a string using the SPDX license expression syntax as
+documented in the `SPDX specification <#spdxpression_>`_, either
+Version 2.2 or a later compatible version.
 
+When used in the ``License-Expression`` field and as a specialization of
+the SPDX license expression definition, a license expression can use the
+following license identifiers:
 
-Core Metadata Specification updates
-===================================
+- Any SPDX-listed license short-form identifiers that are published in the
+  `SPDX License List <#spdxlist_>`_, version 3.15 or any later compatible
+  version. Note that the SPDX working group never removes any license
+  identifiers; instead, they may choose to mark an identifier as "deprecated".
 
-The canonical source for the names and semantics of each of the supported
-metadata fields is the Core Metadata Specification [#cms]_ document.
+- The ``LicenseRef-Public-Domain`` and ``LicenseRef-Proprietary`` strings, to
+  identify licenses that are not included in the SPDX license list.
```

Comment:
I'm wondering if it would make sense to restrict the use together with SPDX identifiers. Not sure if that's really an issue in the end and if I'm just overly worried. What about these cases?
* `LicenseRef-Public-Domain AND CC0-1.0`
* `LicenseRef-Proprietary OR MIT`

Reply:
Hmm, great question! Following this, I went through various combinations, and while some of them (`LicenseRef-Proprietary AND <SPDX-ID>` for proprietary projects with permissive-licensed code) make more sense than others (`LicenseRef-Proprietary OR `LicenseRef-Public-Domain`, though arguably combinations like `AGPL-3.0 OR CC0-1.0` is nearly as nonsensical), I'm not sure the benefits are worth the extra complexity, in the spec and implementations, of micro-managing which IDs can be combined with which and in what ways, at least at the spec level, given the large number of potential ways they can be combined, and with what. But I'm certainly not opposed to it if others feel strongly or have a specific proposal on how this should be handled.

As to the cases:

* `LicenseRef-Public-Domain AND CC0-1.0` could make sense if a project contains code released under a generic public domain declaration, and code later explicitly licensed as CC0. IANAL, but in jurisdictions where releasing a work into the public domain is legally possible, CC0 does that explicitly and LicenseRef-Public-Domain does that implicitly, so it would be equivalent to just CC0, and in jurisdictions where it is not, `LicenseRef-Public-Domain` would be void and just CC0 would apply as well, so it should be effectively equivalent anyway and therefore relatively harmless.
* `LicenseRef-Proprietary OR MIT` This would mean that a project offers a choice of either a FOSS (typically copyleft) or proprietary license (quite common with commercially supported GPL/AGPL packages, like PyQt), and the proprietary license may well grant more rights in some areas than a particularly restrictive copyleft license, but impose restrictions on commercial use/be required to purchase, etc. So unless we drill down into which SPDX IDs are permissable, which I think is way out of scope, I don't think it makes sense to not allow this.

Reply:
> I'm not sure the benefits are worth the extra complexity, in the spec and implementations

Good point! At best this is an edge case anyway. I didn't find a limit in the SPDX spec either. Let's leave it then as is.

Reply:
Sounds good to me!

Reply:
@pombredanne The `OR` case with `LicenseRef-Public-Domain` is self explanatory. I was more concerned about `LicenseRef-Public-Domain AND ...`. What does that actually mean in this context? However, like @CAM-Gerlach mentioned it just indicates that some parts are licensed as `Public-Domain` whereas others aren't.